core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PreferWritePartitioning.java

```diff
@@ -33,6 +47,14 @@ public Result apply(TableWriterNode node, Captures captures, Context context)
             return Result.empty();
         }
 
+        double numberOfPartitions = node.getPreferredPartitioningScheme().get().getPartitioning().getColumns().stream()
+                .mapToDouble(column -> context.getStatsProvider().getStats(node.getSource()).getSymbolStatistics().get(column).getDistinctValuesCount())
+                .max()
+                .orElse(1);
+        if (numberOfPartitions < getPreferredWritePartitioningThreshold(context.getSession())) {
```

Comment:
This should be other way around. If there is lots of partitions, preferred write partitioning should be used.